Auto generate tile edge matching

pierrebai opened the pull request:
Add UI to select the tiles to generate, restart the generation and show tile neighbors options. The UI also gives access to debugging features.

When loading separate tiles, automatically detect which tile can be adjacent to which other tiles by analyzing their edges. This avoids having to specify which edges match which others by hand.  This also enable the capability to load a single image and divide it in tiles.

- Extract the four edges of each tile and fuzzily match it with other edges.
- Make edges size calculation dependent on largest width of tiles.
- Allow applying a filter to the edge images to ignore some differences, for example a background that should be ignored.
- Compare edges then fuzzy image compare to find identical tiles.
- Choose the edge size to be compatible with the small tiles generated from a single-image.
- Limit the comparison range to zero when the edge size is small.
- Make tile rotations be optional.
- Optionally use tile frequencies to choose among the tile options of a cell.

Optimize the wave-function-collapse by using a new Bitmap class.

- Use bitmap unions and intersections to update the grid.
- This is much faster when there are a lots of tiles.
- Change the cell options to be bitmaps to use optimized unions and intersections.
- Replaced the tile up/down/left/right tile options with bitmaps.
- Limit the change propagation to a given radius to speed-up the algorithm.
- But take into account the case where we choose to collapse a cell that was not updated recently, which could lead to choosing a tile that should have been disallowed. We always update the cell options before collapsing it to make sure it has updated options.

(We could also update its neighbors, etc... but we don't need to: if they are already collapsed then they are valid. If they're not
collapsed, they will update themselves later on and if this leads to a contradiction, then we will rewind.)

Add grid history and rewind to avoid restarting from scratch. This is in the `history.js` file. Increase and decrease the rewind distance to get rid of contradictions. That is, by increasing the rewind distance, we can go back far enough to remove the state that invariably lead to a contradiction.

Move the loading of tiles to its own file named `tileLoader.js`. This allows different methods of loading tiles and allow switching between which tiles are used on the fly.

Split the draw and update functions into smaller functions.

Add debug function to help diagnose problems. Add debug logging to show a cell tile and the options of its neighbor, to help debug when there seems to be a contradiction. Also allow showing tile edges and cell tile options. The debugging code is in the `debug.js` file.

shiffman commented:
Wow, this is really quite fantastic! I want to take the time to read over your notes and code more carefully. Because it is very significant update to the codebase I wonder if, instead of overwriting the existing example, it would make sense to create a new `tile-model-optimized` folder (or another appropriately descriptive name) or link to this work in the Community Contributiions list in the README. What do you think?

pierrebai commented:
Yes, no problem, I can move the code to a new folder. Maybe I'll call it hybrid-model since it supports both tiles and single-image sources.

shiffman commented:
That sounds great! Feel free to add any documentation and credits to the README as well!

pierrebai commented:
I don't think I will change the README now, since it would talk about and point to a branch that will disappear if this gets merged. I moved the code the separate folder and also added a few improvements to the tile comparison code while I was at it.

pierrebai commented:
I'm closing this pull request as I created another one with my latest changes, even more optimizations and a new README explaining what I did.
